src/tools/rbd/action/Ubbd.cc

```diff
@@ -0,0 +1,250 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/TextTable.h"
+#include "tools/rbd/Shell.h"
+#include "tools/rbd/Utils.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/scope_exit.hpp>
+#include <iostream>
+#include <sys/stat.h>
+
+#if defined(WITH_RBD_UBBD)
+#include <libubbd.h>
+#endif
+
+namespace rbd {
+namespace action {
+namespace ubbd {
+
+namespace at = argument_types;
+namespace po = boost::program_options;
+
+int execute_list(const po::variables_map &vm,
+                 const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_RBD_UBBD)
+  ubbdd_mgmt_rsp rsp = { 0 };
+  ubbd_list_options list_opts = { .type = UBBD_DEV_TYPE_RBD };
+
+  int r = ubbd_list(&list_opts, &rsp);
+  if (r) {
+    std::cerr << "rbd: ubbd_list failed: " << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  at::Format::Formatter f;
+  r = utils::get_formatter(vm, &f);
+  if (r < 0) {
+    return r;
+  }
+
+  TextTable tbl;
+
+  if (f) {
+    f->open_array_section("devices");
+  } else {
+    tbl.define_column("id", TextTable::LEFT, TextTable::LEFT);
+    tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+    tbl.define_column("namespace", TextTable::LEFT, TextTable::LEFT);
+    tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+    tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+    tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+  }
+
+  for (int i = 0; i < rsp.list.dev_num; i++) {
+    ubbd_info_options info_opts = { .ubbdid = rsp.list.dev_list[i] };
+    ubbdd_mgmt_rsp info_rsp = { 0 };
+    ubbdd_mgmt_rsp_dev_info *mgmt_dev_info = &info_rsp.dev_info;
+    ubbd_dev_info *dev_info = &mgmt_dev_info->dev_info;
+
+    r = ubbd_device_info(&info_opts, &info_rsp);
+    if (r) {
+      std::cerr << "ubbd_device_info for ubbdid=" << info_opts.ubbdid
+                << " failed: " << cpp_strerror(r) << std::endl;
+      return r;
+    }
+
+    if (f) {
+      f->open_object_section("device");
+      f->dump_int("id", mgmt_dev_info->devid);
```

Comment:
Yes, that's always equal. info_opts.ubbdid is the index to find ubbd device, and mgmt_dev_info->devid, is the result returned for this ubbd device. I use mgmt_dev_info->devid because it's the result returned from ubbd demaon, it should be equal with info_opts.ubbdid.

Reply:
Then I'd change the "ubbd_device_info for ubbdid=X failed" error message above to "ubbd_device_info for /dev/ubbdX failed".

Reply:
ok